rflx/model/session.py

```diff
@@ -198,6 +198,105 @@ def _collect_message_field_assignments(
+        """
+
+        def find_identifier(name: ID, expression: expr.Expr) -> bool:
+            """Check if variable with a given identifier exists within an expression."""
+            return bool(
+                expression.findall(lambda e: isinstance(e, expr.Variable) and e.identifier == name)
+            )
+
+        def find_prefix(name: ID, expression: expr.Expr) -> bool:
+            """Check if the prefix of an expression contains a variable with a given identifier."""
+            return bool(
+                expression.findall(
+                    lambda e: isinstance(e, (expr.Attribute, expr.Opaque))
+                    and find_identifier(name, e.prefix)
+                )
+            )
+
+        def substituted(expression: expr.Expr, structure: rty.Structure) -> expr.Expr:
+            """Replace message types with structure of equal name."""
+
+            def subst(expression: expr.Expr) -> expr.Expr:
+                """
+                Substitute type of expression with structure type.
+
+                Replace the type of expression with a structure if
+                it is a message type with the same identifier.
+                """
+                if (
+                    isinstance(expression, (expr.Variable, expr.Call))
+                    and isinstance(expression.type_, rty.Message)
+                    and expression.type_.identifier == structure.identifier
+                ):
+                    expression.type_ = structure
+                return expression
+
+            return expression.substituted(subst)
+
+        def contains_unsupported_feature(name: ID, action: stmt.Statement) -> bool:
+            """Check for features that prevent optimization."""
+            return (
+                # The message is an argument in 'Append.
+                isinstance(action, stmt.Append)
+                and find_identifier(name, action.parameter)
+                # The message is the prefix of an attribute.
+                or (isinstance(action, stmt.AttributeStatement) and action.identifier == name)
+                # The message is the prefix of an assignment target.
+                or (isinstance(action, stmt.Assignment) and find_prefix(name, action.expression))
+                # A field of the message is assigned.
+                or (isinstance(action, stmt.MessageFieldAssignment) and action.message == name)
+                # The message is an argument in 'Extend.
+                or (
+                    isinstance(action, stmt.Extend)
+                    and any(p for p in action.parameters if find_identifier(name, p))
+                )
+                # Assignment to the message that is not a call.
+                or (
+                    isinstance(action, stmt.VariableAssignment)
+                    and action.identifier == name
+                    and not isinstance(action.expression, expr.Call)
+                )
```

Comment:
I have the feeling that these comments are not really helpful. In some cases, I'm also not sure if they say the right thing, e.g., in `The message is the prefix of an assignment target.` IMO the assignment target would be `action.identifier` and not `action.expression`.

Reply:
I have removed the comments and refactored the whole statement as it contained some redundancies.